### PR TITLE
Issue 2591: Security documentation

### LINKED ISSUE #2591: Pravega Security: Update Pravega wiki with security documentation
**Problem description**
Update Pravega wiki with security documentation

**Problem location**
Pravega wiki.

**Suggestions for an improvement**
Update the documentation.

### PR DESCRIPTION
**Pull request by arvindkandhare**
**Change log description**
Documents Pravega security and TLS and encryption feature.

**Purpose of the change**
Fixes #2591.
 
**What the code does**
Documents Pravega security and TLS and encryption feature.

**How to verify it**
Read the documentation.

### CODE REVIEW

File: documentation/src/docs/auth/encryption_dependencies.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<!--
+Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
```
Comment (fpj) [inline documentation/src/docs/auth/encryption_dependencies.md]: Empty document.

File: documentation/src/docs/deployment/manual-install.md [hunk 2]
```diff
@@ -153,3 +155,5 @@ Once the configuration changes have been made you can start the segment store wi
 ```
 bin/pravega-segmentstore
 ```
+## Runing Pravega Controller and Segment Store with security enabled
+`TODO`
```
Comment (fpj) [inline documentation/src/docs/deployment/manual-install.md]: Todo.

File: documentation/src/docs/auth/auth-plugin.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+# Implement Pravega Authentication/Authorization plugin
+
+This guide describes the authentication/authorization plugin model for Pravega.
+
+## Pravega auth interface
+
```
Comment (fpj) [inline documentation/src/docs/auth/auth-plugin.md]: Empty section.

File: documentation/src/docs/index.md [hunk 4]
```diff
@@ -49,6 +51,12 @@ You can find a list of frequently asked questions [here](faq.md).
 
 Pravega can be installed locally or in a distributed environment. The installation and deployment of Pravega is covered in the [Running Pravega](deployment/deployment.md) guide.
 
+## Pravega Security, Role based access control and TLS
+Pravega supports encryption of all communication channels and pluggable role based access control.
+Find more details here:
+1. [TLS](security/pravega-security-encryption.md)
+2. [Authorization, authentication and RBAC](security/pravega-security-authorization-authentication.md)
+
 ## Support
```
Comment (fpj) [inline documentation/src/docs/index.md]: I think a `security@pravega.io` address would be a much better way to communicate with us on security issues. We should set it up.

File: documentation/src/docs/auth/encryption.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<!--
+Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
```
Comment (fpj) [inline documentation/src/docs/auth/encryption.md]: Empty document.

File: documentation/src/docs/security/pravega-security-encryption.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+### Certificate management
+Pravega expects administrator user to create and manage certificate creation, deployment and management. 
+Pravega provides different configuration parameters where certificates for different communication channels can be specified.
+### Encrypted data flow between Pravega client and Pravega controller and segmentstore
+Pravega uses same cert to interact with both controller and segmentstore. The certificates need to be mentioned specifically on the client and the server.
+These certificates are not loaded from the trust store.
+### Encrypted data flow between Pravega and tier-1 (Apache Bookkeeper)
+Apache bookkeeper supports JKS based truststore. Segmentstore uses JKS based truststore to interact with it. The configurations
+#### Segmentstore and bookkeeper
+Pravega Segmentstore uses Apache bookkeeper as Tier-1 storage.
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-encryption.md]: `Tier-x` is not being written consistently, please use `Tier 1` and `Tier 2` when referring to Pravega storage tiers, and `tier` small font when talking about tiers in general.

File: documentation/src/docs/auth/client-auth.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+# Client Auth interface
+Pravega client can access Pravega APIs through `grpc`. Some of the admin APIs can be accessed via `REST` API. 
+The authorization/authentication API and plugin works for both these interfaces.
+
+## grpc client auth interface
+In case more than one plugin exists, a client selects its auth handler by setting a grpc header with a name "method". 
+This is done by implementing [Credentials](https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/stream/impl/Credentials.java) interface.
```
Comment (fpj) [inline documentation/src/docs/auth/client-auth.md]: Would it make sense to include code snippets showing a simple example?

File: documentation/src/docs/security/pravega-security-authorization-authentication.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+Having different levels of access to the data. This feature is crucial to supprt multi tenancy.
+
+Following key features are added as part of security implementation:
+1. Administrators can enable encryption for different communication channels using TLS.
+2. A Unix-like permission system which controls which users can access which part of the data and the levels of access.
+3. Drop in implementations of the authorization/authentication APIs can be dropped in and loaded dynamically.
+Multiple such implementations can co-exist and different users can use different plugins.
+4. All the other compnents that are deployed with Pravega (e.g. Bookkeeper, Zookeeper etc) can be deployed securely with TLS.
+
+[PDP-23](https://github.com/pravega/pravega/wiki/PDP-23:-Pravega-security----encryption-and-Role-Based-Access-Control)
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-authorization-authentication.md]: Even if we don't do it here, we should eventually refine the text in the PDP and move to our final documentation.

File: documentation/src/docs/security/pravega-security-encryption.md [hunk 9]
```diff
@@ -18,10 +18,10 @@ Pravega provides different configuration parameters where certificates for diffe
 ### Encrypted data flow between Pravega client and Pravega controller and segmentstore
 Pravega uses same cert to interact with both controller and segmentstore. The certificates need to be mentioned specifically on the client and the server.
 These certificates are not loaded from the trust store.
-### Encrypted data flow between Pravega and tier-1 (Apache Bookkeeper)
+### Encrypted data flow between Pravega and tier 1 (Apache Bookkeeper)
 Apache bookkeeper supports JKS based truststore. Segmentstore uses JKS based truststore to interact with it. The configurations
 #### Segmentstore and bookkeeper
-Pravega Segmentstore uses Apache bookkeeper as Tier-1 storage. 
+Pravega Segmentstore uses Apache bookkeeper as Tier 1 storage.
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-encryption.md]: This section seems to be incomplete.

File: documentation/src/docs/security/pravega-security-encryption.md [hunk 10]
```diff
@@ -18,10 +18,10 @@ Pravega provides different configuration parameters where certificates for diffe
 ### Encrypted data flow between Pravega client and Pravega controller and segmentstore
 Pravega uses same cert to interact with both controller and segmentstore. The certificates need to be mentioned specifically on the client and the server.
 These certificates are not loaded from the trust store.
-### Encrypted data flow between Pravega and tier-1 (Apache Bookkeeper)
+### Encrypted data flow between Pravega and tier 1 (Apache Bookkeeper)
 Apache bookkeeper supports JKS based truststore. Segmentstore uses JKS based truststore to interact with it. The configurations
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-encryption.md]: This sentence is incomplete.

File: documentation/src/docs/security/pravega-security-configurations.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
+<!--
+Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+#Pravega Security Configurations
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-configurations.md]: If you click on the "View" button above, you'll see that this table is not properly formatting.

Comment (arvindkandhare) [reply documentation/src/docs/security/pravega-security-configurations.md]: Fixed.

File: documentation/src/docs/security/pravega-security-encryption.md (hunk 10 above)
Comment (arvindkandhare) [reply documentation/src/docs/security/pravega-security-encryption.md]: Fixed.

File: documentation/src/docs/security/pravega-security-encryption.md (hunk 9 above)
Comment (arvindkandhare) [reply documentation/src/docs/security/pravega-security-encryption.md]: Fixed.

File: documentation/src/docs/auth/auth-plugin.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
+<!--
+Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+# Implement Pravega Authentication/Authorization plugin
```
Comment (Aparnarr) [inline documentation/src/docs/auth/auth-plugin.md]: Can i change as
Implementation of Pravega Authentication/Authorization plugin

File: documentation/src/docs/auth/auth-plugin.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+This guide describes the authentication/authorization plugin model for Pravega.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/auth-plugin.md]: This guide describes in detail the authentication/authorization plugin model for Pravega

File: documentation/src/docs/auth/auth-plugin.md [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+## Pravega auth interface
```
Comment (Aparnarr) [inline documentation/src/docs/auth/auth-plugin.md]: Pravega _auth_ interface

File: documentation/src/docs/auth/auth-plugin.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+The custom implementation is expected to implement the [AuthHandler](https://github.com/pravega/pravega/blob/master/shared/authplugin/src/main/java/io/pravega/auth/AuthHandler.java) interface.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/auth-plugin.md]: The custom implementation performs the implementation of the [AuthHandler](https://github.com/pravega/pravega/blob/master/shared/authplugin/src/main/java/io/pravega/auth/AuthHandler.java) interface.

File: documentation/src/docs/auth/auth-plugin.md [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+## Dynamically loading auth implementations
```
Comment (Aparnarr) [inline documentation/src/docs/auth/auth-plugin.md]: Dynamic loading of _auth_ implementations.

File: documentation/src/docs/auth/auth-plugin.md [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+Administrator users can implement their own authorization/authentication plugins. Multiple such plugins can exist together.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/auth-plugin.md]: Administrators and  users are allowed to  implement their own Authorization/Authentication plugins. Multiple plugins of such kind exists together.

File: documentation/src/docs/auth/auth-plugin.md [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+The plugin implementation follows the [Java Service Loader](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html) approach.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/auth-plugin.md]: The  implementation of plugin  follows the  [Java Service Loader](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html)  approach.

File: documentation/src/docs/auth/client-auth.md [hunk 19]
```diff
@@ -0,0 +1,35 @@
+<!--
+Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+# Client Auth interface
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: Please mention a consistent format for: _auth_ Interface or _Auth_ Interface
Client _Auth_ Interface

File: documentation/src/docs/auth/client-auth.md [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+Pravega client can access Pravega APIs through `grpc`. Some of the admin APIs can be accessed via `REST` API. 
+The authorization/authentication API and plugin works for both these interfaces.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: The Authorization/Authentication API and plugin works for both these interfaces.

File: documentation/src/docs/auth/client-auth.md [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+## grpc client auth interface
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: _auth_ interface

File: documentation/src/docs/auth/client-auth.md [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+In case more than one plugin exists, a client selects its auth handler by setting a grpc header with a name "method".
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: If multiple plugin exists then a client selects its _auth_ handler by setting a `grpc` header naming it as  `method`.

File: documentation/src/docs/auth/client-auth.md (hunk 7 above)
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: This is performed by implementing [Credentials](https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/stream/impl/Credentials.java) interface by  passing through the [ClientConfig](https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/ClientConfig.java) object to the Pravega client.

File: documentation/src/docs/auth/client-auth.md [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+# Client Auth interface
+Pravega client can access Pravega APIs through `grpc`. Some of the admin APIs can be accessed via `REST` API. 
+The authorization/authentication API and plugin works for both these interfaces.
+
+## grpc client auth interface
+In case more than one plugin exists, a client selects its auth handler by setting a grpc header with a name "method". 
+This is done by implementing [Credentials](https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/stream/impl/Credentials.java) interface.
+ This is passed through the [ClientConfig](https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/ClientConfig.java) object to the Pravega client.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: please see above comment

File: documentation/src/docs/auth/client-auth.md [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+The authentication related parameters are passed through custom grpc headers. These are extracted through grpc interceptors and passed on to the specific auth plugin.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: The parameters for authentication are passed through the custom `grpc` headers which are extracted through `grpc` interceptors and passed on to the specific _auth_ plugin.

File: documentation/src/docs/auth/client-auth.md [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+This plugin is identified by the `method` header.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: The plugin is identified by the 'method' header.

File: documentation/src/docs/auth/client-auth.md [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+
+## Dynamic extraction of the auth parameters on the client
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: _auth_

File: documentation/src/docs/auth/client-auth.md [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+The parameters can also be extracted dynamically from the system properties OR environment variables.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: Dynamic extraction of parameters is also possible using the system properties or environment variables.

File: documentation/src/docs/auth/client-auth.md [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+Here is the order of preference in descending order
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: The order of preference  is given below: 
Arvind! could you provide some clarity on mentioning  ( descending order!@)

File: documentation/src/docs/auth/client-auth.md [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+
+1. User explicitly provides a credential object through the API. This overrides any other settings.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: User explicitly provides a credential object through the API. This results in  overriding other settings.

File: documentation/src/docs/auth/client-auth.md [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+2. System properties: System properties are defined in the format: `pravega.client.auth.*`
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: _System Properties_: System properties are defined in the format: `pravega.client.auth.*`

File: documentation/src/docs/auth/client-auth.md [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+3. Environment variables. Environment variables are defined under the format `pravega_client_auth_*`
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: _Environment variables: Environment variables are defined in the format `pravega_client_auth_*`

File: documentation/src/docs/auth/client-auth.md [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+4. In case of option 2 and 3, the caller can decide whether the class needs to be loaded dynamically by setting property `pravega.client.auth.loadDynamic` to true.
+ 
+# REST client auth interface
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: _auth_

File: documentation/src/docs/auth/client-auth.md [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+The REST client to access Pravea API uses the similar approach. The custom auth parameters are sent as part of the `Authorization` HTTP header.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: The REST client in order to access Pravega API uses the similar approach. The custom _auth_ parameters are sent as part of the `Authorization` HTTP header.

File: documentation/src/docs/auth/client-auth.md [hunk 34]
```diff
@@ -0,0 +1,35 @@
...
+
+The REST server implementation on Pravega Controller extracts these headers, passes it to valid auth plugin implementation
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: The REST server implementation on Pravega Controller extracts these headers, passes it to the valid _auth_ plugin implementation.

File: documentation/src/docs/auth/client-auth.md [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+and then resumes with the call if the authentication and authorization matches the intended access pattern.
```
Comment (Aparnarr) [inline documentation/src/docs/auth/client-auth.md]: Can it be as follows:
It then resumes, if the authentication and authorization matches the intended access pattern.

File: documentation/src/docs/deployment/manual-install.md [hunk 36]
```diff
@@ -153,3 +155,6 @@ Once the configuration changes have been made you can start the segment store wi
 ```
 bin/pravega-segmentstore
 ```
+## Runing Pravega Controller and Segment Store with security enabled
+Here are the details for configurations for [Pravega controller](../security/pravega-security-configurations.md#Pravega controller) and [Pravega Segmentstore](../security/pravega-security-configurations.md#Pravega Segmentstore).
```
Comment (Aparnarr) [inline documentation/src/docs/deployment/manual-install.md]: Please provide links

File: documentation/src/docs/security/pravega-security-configurations.md [hunk 37]
```diff
@@ -0,0 +1,50 @@
...
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+# Pravega Security Configurations
+
+The following is the list of configuration parameters for different Pravega components.
+
+## Pravega Segmentstore
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-configurations.md]: `Segment store` two words, please.

File: documentation/src/docs/security/pravega-security-encryption.md [hunk 38]
```diff
@@ -0,0 +1,35 @@
...
+**Note:** These certificates are not loaded from the truststore.
+
+### Encrypted data flow between Pravega and Tier 1 (Apache Bookkeeper)
+Pravega segmentstore uses Apache Bookkeeper as Tier 1 storage. Apache Bookkeeper supports JKS based truststore. Segmentstore uses JKS based truststore to interact with it. 
+The configurations can be found [here](pravega-security-configurations.md#pravega-segmentstore).
+
+### Encrypted access to Apache Zookeeper
+This implementation is still in [progress](https://github.com/pravega/pravega/issues/2034). 
+
+## Encryption of data at REST
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-encryption.md]: I don't understand `data at REST`. The capitalized `REST` seems to imply the controller REST API, but I don't know whether this is referring to something else.

File: documentation/src/docs/security/pravega-security-encryption.md [hunk 39]
```diff
@@ -0,0 +1,35 @@
...
+### Encryption of data in Tier 2
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-encryption.md]: Swap `Encryption of data in Tier 2` with ``Encryption of data in Tier 1`. Let's follow the numerical order, 1 then 2.

File: documentation/src/docs/security/pravega-security-encryption.md [hunk 40]
```diff
@@ -0,0 +1,35 @@
...
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+# Pravega Encryption
+## Encryption of data in flight 
+Pravega ensures that all the data in flight can be passed by applying encrytion.
```
Comment (fpj) [inline documentation/src/docs/security/pravega-security-encryption.md]: Typo in `encrytion`.

**Comment by arvindkandhare** [comment]
Closing this PR as discussed with @Aparnarr .

**Pull request closed without merging**